src/order/partition/finpartition.lean

```diff
@@ -0,0 +1,456 @@
+  (hf : f i ≠ ⊥) (hij : f i ≤ f j) :
+  i = j :=
+begin
+  classical,
+  by_contra,
+  exact hf (disjoint_self.1 $ (hs hi hj h).mono_right hij),
+end
+
+lemma mk_mem_product {β : Type*} {s : finset α} {t : finset β} {a : α} {b : β} (ha : a ∈ s)
+  (hb : b ∈ t) :
+  (a, b) ∈ s.product t :=
+mem_product.2 ⟨ha, hb⟩
+
+/-- A finite partition of `a : α` is a pairwise disjoint finite set of elements whose supremum is
+`a`. We forbid `⊥` as a part. -/
+@[ext, derive decidable_eq] structure finpartition {α : Type*} [lattice α] [order_bot α] (a : α) :=
+(parts : finset α)
+(sup_indep : parts.sup_indep id)
+(sup_parts : parts.sup id = a)
+(not_bot_mem : ⊥ ∉ parts)
+
+attribute [protected] finpartition.sup_indep
+
+namespace finpartition
+section lattice
+variables [lattice α] [order_bot α]
+
+/-- A `finpartition` constructor which does not insist on `⊥` not being a part. -/
+@[simps] def of_erase [decidable_eq α] {a : α} (parts : finset α)
+  (sup_indep : parts.sup_indep id) (sup_parts : parts.sup id = a) :
+  finpartition a :=
+{ parts := parts.erase ⊥,
+  sup_indep := sup_indep.subset (erase_subset _ _),
+  sup_parts := (sup_erase_bot _).trans sup_parts,
+  not_bot_mem := not_mem_erase _ _ }
+
+/-- A `finpartition` constructor from a bigger existing finpartition. -/
+@[simps] def of_subset {a b : α} (P : finpartition a) {parts : finset α}
+  (subset : parts ⊆ P.parts) (sup_parts : parts.sup id = b) :
+  finpartition b :=
+{ parts := parts,
+  sup_indep := P.sup_indep.subset subset,
+  sup_parts := sup_parts,
+  not_bot_mem := λ h, P.not_bot_mem (subset h) }
+
+/-- Changes the type of a finpartition to an equal one. -/
+@[simps] def copy {a b : α} (P : finpartition a) (h : a = b) : finpartition b :=
+{ parts := P.parts,
+  sup_indep := P.sup_indep,
+  sup_parts := h ▸ P.sup_parts,
+  not_bot_mem := P.not_bot_mem }
+
+variables (α)
+
+/-- The empty finpartition. -/
+@[simps] protected def empty : finpartition (⊥ : α) :=
+{ parts := ∅,
+  sup_indep := sup_indep_empty _,
+  sup_parts := finset.sup_empty,
+  not_bot_mem := not_mem_empty ⊥ }
+
+instance : inhabited (finpartition (⊥ : α)) := ⟨finpartition.empty α⟩
+
+variables {α} {a : α}
+
+/-- The finpartition in one part, aka indiscrete finpartition. -/
+@[simps] def indiscrete (ha : a ≠ ⊥) : finpartition a :=
+{ parts := {a},
+  sup_indep := sup_indep_singleton _ _,
+  sup_parts := finset.sup_singleton,
+  not_bot_mem := λ h, ha (mem_singleton.1 h).symm }
+
+variables (P : finpartition a)
+
+protected lemma le {b : α} (hb : b ∈ P.parts) : b ≤ a := (le_sup hb).trans P.sup_parts.le
+
+lemma ne_bot {b : α} (hb : b ∈ P.parts) : b ≠ ⊥ := λ h, P.not_bot_mem $ h.subst hb
+
+protected lemma disjoint : (P.parts : set α).pairwise_disjoint id := P.sup_indep.pairwise_disjoint
+
+lemma eq_empty (P : finpartition (⊥ : α)) : P = finpartition.empty α :=
+begin
+  ext a,
+  exact iff_of_false (λ h, P.ne_bot h $ le_bot_iff.1 $ P.le h) (not_mem_empty a),
+end
```

Comment:
This statement would be better as `unique (finpartition (⊥ : α))`

Reply:
That's what it is if you go below. I can inline it.

Reply:
Ah - which is to say, I think you should inline this proof into the `unique` instance below.